staging/src/k8s.io/kubelet/pkg/cri/streaming/portforward/httpstream.go

```diff
@@ -256,6 +256,12 @@ func (h *httpStreamHandler) portForward(p *httpStreamPair) {
 		msg := fmt.Errorf("error forwarding port %d to pod %s, uid %v: %v", port, h.pod, h.uid, err)
 		utilruntime.HandleError(msg)
 		fmt.Fprint(p.errorStream, msg.Error())
+		// receiving an error from a port-forward operation indicates a problem
```

Comment:
we have two defers
```
	defer p.dataStream.Close()
	defer p.errorStream.Close()
```

can it be a problem to Close over a Reseted stream?

Reply:
alternatively, should we unconditionally Reset() when we're done, rather than Close(), to fully close/drain both directions of both streams?

Reply:
if you unconditionally reset() you don't leave much time to the other side to gracefully terminate, no?

A ----> Close(). --> B io.EOF() do my thing and send Close() back

Reply:
I thought all of that completed before PortForward returned... does it not?

Reply:
I don't expect problems with Reset followed by a Close. Looking at [spdy implementation of both of these methods](https://github.com/moby/spdystream/blob/77eb080123d208697674a07b74ceaf94c98bee8b/stream.go#L171-L186) the main difference is like Antonio says, Reset sends a RST frame, whereas Close sends an empty frame indicating FIN. 

I've done several tests locally, since we're also have a similar situation in client-go (see https://github.com/kubernetes/kubernetes/pull/128319/) reseting streams (specifically `dataStream.Reset()` unconditionally (this is coming from https://github.com/kubernetes/kubernetes/pull/126718) and then closing them in defer in the goroutine responsible for sending data to server (https://github.com/kubernetes/kubernetes/blob/8b063a6a08bc27e91a38c8ec6099cb6782e16a08/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L394). Everything is working as expected and no errors are reported during Reset+Close operation.

Reply:
> the main difference is like Antonio says, Reset sends a RST frame, whereas Close sends an empty frame indicating FIN

is it incorrect to send more stuff (like an empty frame from a Close()) after having sent a RST frame?

Reply:
I'll defer to @aojea he's the networking expert :sweat_smile:

Reply:
let's see the spdystream implementation

Reset()
```go
// Reset sends a reset frame, putting the stream into the fully closed state.
func (s *Stream) Reset() error {
	s.conn.removeStream(s)
	return s.resetStream()
}

func (s *Stream) resetStream() error {
	// Always call closeRemoteChannels, even if s.finished is already true.
	// This makes it so that stream.Close() followed by stream.Reset() allows
	// stream.Read() to unblock.
	s.closeRemoteChannels()

	s.finishLock.Lock()
	if s.finished {
		s.finishLock.Unlock()
		return nil
	}
	s.finished = true
	s.finishLock.Unlock()

	resetFrame := &spdy.RstStreamFrame{
		StreamId: s.streamId,
		Status:   spdy.Cancel,
	}
	return s.conn.framer.WriteFrame(resetFrame)
}

```

it calls  closeRemoteChannels()
```

func (s *Stream) closeRemoteChannels() {
	s.closeLock.Lock()
	defer s.closeLock.Unlock()
	select {
	case <-s.closeChan:
	default:
		close(s.closeChan)
	}
}
```

and then Close()
```
// Close closes the stream by sending an empty data frame with the
// finish flag set, indicating this side is finished with the stream.
func (s *Stream) Close() error {
	select {
	case <-s.closeChan:
		// Stream is now fully closed
		s.conn.removeStream(s)
	default:
		break
	}
	return s.WriteData([]byte{}, true)
}
```


Close() bails out before trying to write

It is safe

Reply:
it seems this is safe

/lgtm

I've should done that before asking the question, sorry about that